source/common/upstream/load_balancer_impl.h

```diff
@@ -51,8 +55,18 @@ class LoadBalancerBase {
   void calculateZonePercentage(const std::vector<std::vector<HostPtr>>& hosts_per_zone,
                                uint64_t* ret);
 
+  /**
+   * Regenerate zone aware routing structures for fast decisions on upstream zone selection.
+   */
+  void regenerateZoneRoutingStructures();
+
   const HostSet& host_set_;
   const HostSet* local_host_set_;
+
+  bool early_exit_zone_routing_;
```

Comment:
Actually, early_exit_zone_routing_ and route_directly_ are mutually exclusive. Combine these into a state_ enum which has something like [NoZoneRouting, ZoneDirect, ZoneResidual] or something. You can have cleaner logic and better asserts this way.